Add autoConfigure

pete-thompson opened the pull request:
I've added a method to allow quick configuration of the midpoint and noise level based on taking readings. I'm finding that in my circuit the 5V line isn't always 5v (it's 5.1 when connected to my Mac USB), hence the midpoint isn't where it's expected and the easiest way for me to configure is to capture actual ADC values with no current running.

pete-thompson commented:
BTW - feel free to remove my name from author and update version! I hadn't realised you be so quick to respond to things and thought I'd end up leaving this on Github for a while...

I picked version 0.2.0 since I've added API functions

RobTillaart commented:
The autoconfigure code looks OK, some questions:

The code assumes ```midpoint = (minread + maxread) / 2;```   
Makes good sense.
Would it be better to take the average of all readings as midpoint?
In your proposed code a single glitch could influence the midpoint quite a bit.

opinion?

RobTillaart commented:
You co authored the code, so credits where credits are due 👍 
Please check the library.json file to mention you as coauthor too

pete-thompson commented:
Thanks Rob. Using the mean sounds a good plan - I'll have a look at trying that out one evening this week (hopefully!). I'm also tempted to try using standard deviation to calculate a value for noise, although that will involve more experimentation and some research into what standard deviation is for a simple noise signal (I found one article suggesting 6-8 * sd for noise, which sounds plausible but I'll need to play...). Might be utter overkill really!

RobTillaart commented:
Calculating the std deviation for the noise would be good science, however it means extra code compared to core functionality of the lib. Especially for smaller platforms this is a point of attention.

The current peak to peak approach is fast and lightweight, so I'm looking forward to the results of your experiments.
If there is little difference I prefer the current pragmatic approach as it uses less code (= assumption)

pete-thompson commented:
Good point - I'm spoiled coming from a background of building for servers rather than microcontrollers :) I'll play and see..

pete-thompson commented:
Hi Rob. I ran a test and captured 6,000 readings with no current flowing. The values generally fell nicely within 3SDs of the mean, but there were quite a few outliers (more than 250) - so you're right and the peak to peak approach for midpoint is questionable. I've updated the logic so that I calculate the mean value instead. I also decided that sticking with the data sheet value of 21mV for noise was a better plan than using the peak to peak values - in my test 3 standard deviations was almost exactly 21mV. I agree that calculating the SD during a test is overkill - the data sheet value is probably safe unless you have a defective component...

Sadly, all this has gotten me wondering if I need to worry about noise when taking AC readings. If I take lots of DC readings the noise will cancel itself out over time, but the same isn't true for AC. I'll run some tests and have a deeper think about it soon - although I think I might be overthinking all this!

pete-thompson commented on README.md:
```diff
@@ -4,20 +4,20 @@ Current Sensor - 5A, 20A, 30A
 
 ## Description
 
-The ACS712 is a chip to measure current, both AC or DC. The chip has an 
```
Sorry - looks like my editor has stripped trailing spaces

RobTillaart replied:
no problem

RobTillaart commented:
Thanks, but I'm sorry to say that the code has two possible showstoppers in it

**overflow**
The autoMIdpoint can run about a minute max, 
Although **total** is not expected to overflow normally an ESP32 @240 Mhz will make more than 5.7 million samples in such loop (tested). As the value of one sample = ~2048  (50% of default 12 bit mode) ==> 5.7 million * 2048  > MAX_UINT32_T
==> should be prevented in the ```while``` loop.

**samples**
The code has a bug as it initializes **samples** to 1 which is incorrect.

If the user passes **timeMillis = 0** as parameter or a value small enough, then the ``` while``` loop will not be executed. **total** will stay zero and the result is that the midpoint will be 0, which is 100% incorrect.

If the user passes a small value and the loop executes only once, **samples** will be 2 and the midpoint will be wrong too.
The more iterations of loop the better it will become.

There are different solutions and I propose the following:
- use a minimum nr of samples. ==> one can enforce a decent minimum (preferred) 
- exit the loop if total approaches an overflow condition

```cpp 
void ACS712::autoMidPoint(uint16_t timeMillis)
{
  uint32_t startTime = millis();
  uint32_t total = 0;
  uint32_t samples = 0;
  while ( ((millis() - startTime) < timeMillis) && (samples < 4) && (total < 0xFF000000) )
  { 
    total += analogRead(_pin);
    samples ++;
  }
  _midPoint = (total + samples/2) / samples;     // better int rounding;  no big issue as this could only give +1.
}
```

or maybe enforce at least one sample 
```cpp 
void ACS712::autoMidPoint(uint16_t timeMillis)
{
  uint32_t startTime = millis();
  uint32_t total = analogRead(_pin);
  uint32_t samples = 1;
  while ( ((millis() - startTime) < timeMillis) && (total < 0xFF000000) )
  { 
    total += analogRead(_pin);
    samples ++;
  }
  _midPoint = total / samples; 
```

maybe it is better to enforce a minimum timeMillis so that for AC sampling one het at least a full sinus of (50/60 Hz)
Opinion?

pete-thompson commented:
Thanks Rob! The samples=1 mistake was a holdover from an earlier version where I attempted to calculate mean around the initial reading to limit overflow possibilities (I never got it to work because of signed/unsigned weirdness!). Anyway, fixed that...

I've gone with a combination of your suggestions and had a little more of a think about calculating midpoint if there's an AC current flowing (my original assumption was that there was no current). I've modified the logic so that it takes readings for 2 cycles of AC, and pauses for a microsecond between readings. That way I can be sure I get a maximum of 40,000 readings and therefore won't hit an overflow. I'm getting around 2,800 samples on my ESP32 with this logic, which seems plenty to get a reasonable mean.

Hopefully I've managed to do something that makes sense!

RobTillaart commented:
Looks much more robust, don't see any failing scenario, so I will merge asap.

An analogRead on an UNO takes ~120 micros so that would give enough samples too.

RobTillaart commented:
Released 0.2.0

pete-thompson commented:
Excellent. Thanks Rob
